### PR TITLE
FFmpeg: Misc improvements, use MD5 muxer and add CUDA/NVDEC decoders

### PR DESCRIPTION
**Pull request by Kwiboo**
This PR include misc improvements and refactorings to simplify building the FFmpeg command line.

- Rename and refactor of `utils.run_pipe_command_with_std_output()`
- Use shared `@lru_cache` for ffmpeg commands used in `check()`
- Fix version detect of `<master>`.`<minor>` releases
- Separate handling of `-hw_output_format` and `-hwaccel_output_format`
- Convert `output_format_to_ffformat()` to `hw_download_mapping` property
- Refactor to build command line in `decode()` using a string list
- Validate and inform what codec to use using `-codec`
- Use `md5` muxer when supported to avoid always having to save output files
- Use single-threaded decoding, `fluster` support running multiple jobs
- Rename V4L2 mem2mem decoders
- Add VDPAU VP9 decoder
- Add CUDA/NVDEC decoders

This was tested using FFmpeg `4.2.1` and `7.0.1` along with GStreamer `1.24.5`.

### CODE REVIEW

**Comment by ceyusa** [comment]
I've tested these patches wit ffmpeg/vulkan decoding and it worked ok.

File: fluster/decoders/ffmpeg.py [hunk 1]
```diff
@@ -119,38 +72,111 @@ def decode(
...
+            Codec.VP9: "vp9",
+            Codec.AV1: "av1",
+        }
+        if self.codec not in codec_mapping:
+            return False
+        self.ffmpeg_codec = codec_mapping[self.codec]
+
+        # Get ffmpeg version
+        output = _run_ffmpeg_command(self.binary, "-version", verbose=verbose)
+        version = re.search(r"ffmpeg version (\d+)\.(\d+)(?:\.(\d+))?", output)
```
Comment (ceyusa) [inline fluster/decoders/ffmpeg.py:158]: this re doesn't match when ffmpeg is compiled from git (N-git-describe)

Comment (Kwiboo) [reply fluster/decoders/ffmpeg.py]: I have adjusted the pattern to also match builds from `release/` branches (`n`-prefix) and changed the version check to presume that `-fps_mode` can be used when no version can be determined.

**Comment by Kwiboo** [comment]
Updated to also work with older releases, e.g. `release/2.8`.
- Update version regex to also match builds from `release/` branch
- Drop use of `repeat`-flag in `-loglevel`
- Change to use `-formats` instead of `-muxers`

File: fluster/decoders/gstreamer.py [hunk 2]
```diff
@@ -111,13 +111,10 @@ def gen_pipeline(
             self.cmd, input_filepath, self.decoder_bin, self.caps, self.sink, output
         )
 
-    def parse_videocodectestsink_md5sum(self, data: List[str], verbose: bool) -> str:
+    def parse_videocodectestsink_md5sum(self, data: List[str]) -> str:
```
Comment (rsanchez87) [inline fluster/decoders/gstreamer.py:114]: What do you think about these changes (L117-L137)?

    def parse_videocodectestsink_md5sum(self, data: List[str]) -> str:
        """Parse the MD5 sum out of commandline output produced when using
        videocodectestsink."""
        # Define the pattern to search for in each line
        pattern = "conformance/checksum, checksum-type=(string)MD5, checksum=(string)"
        # Iterate over each line in the data
        for line in data:
            # Partition the line into three parts: before the pattern, the pattern, and after the pattern
            before, match, after = line.partition(pattern)
            # If the pattern is found (match is not empty)
            if match:
                # Partition the remaining part to find the checksum up to the first ';'
                checksum, _, _ = after.partition(";")
                # If a valid checksum is found
                if checksum:
                    return checksum

        raise Exception("No MD5 found in the program trace.")

**Review by rsanchez87 — CHANGES_REQUESTED** [review]

File: fluster/decoders/gstreamer.py (hunk 2 above)
Comment (rsanchez87) [reply fluster/decoders/gstreamer.py:114]: I'm going to change in another PR

Comment (Kwiboo) [reply fluster/decoders/gstreamer.py:114]: Thanks, and I agree that those changes are probably better to do in a separate PR, my changes where just bare minimum to move the verbose print to the utils function, where I think it belongs.

**Review by rsanchez87 — APPROVED** [review]

**Comment by mdimopoulos** [comment]
Thank you all for your awesome contributions, it's been a pleasure!
Some test results on my Intel machine actually improved with these changes.

**Comment by Kwiboo** [comment]
@mdimopoulos Thanks!

I am just guessing that the improved score may have to do with the change to use `-threads 1`.

Prior to the use of `-threads 1` I noticed several warnings of too many surfaces being used (more than 32) due to high cpu core count on my machine/server. The reduced / explicit `thread_count` help limit the number of surfaces some hwaccels uses.

With CUDA/NVDEC some tests still cause such issue, and it report that `33` surfaces is being used, and that is more than the supported `32`. That was `32` for the reference and reordering frames, and then it added an extra surface for each `thread_count` making it `33` with ´-threads 1`, something that may need to be fixed/changed in FFmpeg for such hwaccels.

**Comment by mdimopoulos** [comment]
Great, appreciate the info, i hadn't noticed the warnings myself 🙈.

Regarding
> With CUDA/NVDEC some tests still cause such issue, and it report that `33` surfaces is being used, and that is more than the supported `32`. That was `32` for the reference and reordering frames, and then it added an extra surface for each `thread_count` making it `33` with ´-threads 1`, something that may need to be fixed/changed in FFmpeg for such hwaccels.

is it an issue because of FFmpeg itself or are you referring to the definition of FFmpeg decoder within fluster?

**Comment by Kwiboo** [comment]
> is it an issue because of FFmpeg itself or are you referring to the definition of FFmpeg decoder within fluster?

I see this purely as a FFmpeg issue, I think there should not be any need to add an extra surface when a single thread is being used, but have not tried to look into that.

My main goal with this PR was to prepare for #179, adding CUDA/NVDEC was more of a bonus and it also helped me run fluster tests directly on my server equipped with a NVIDIA Quadro T400 card.

**Comment by mdimopoulos** [comment]
Great, thanks again!
Also, good to know that tests run fine with the pro line of Nvidia GPUs ;)

**Pull request merged**